azkaban-common/src/main/java/azkaban/utils/YarnUtils.java

```diff
@@ -107,17 +112,32 @@ public static List<ApplicationReport> getAllAliveAppReportsByExecIDs(final YarnC
 
 
   /**
-   * Uses YarnClient to kill the jobs one by one
+   * Uses YarnClient to kill the jobs one by one, each kill has a timeout
    */
   public static void killAllAppsOnCluster(YarnClient yarnClient, Set<String> applicationIDs,
       Logger log) {
     log.info(String.format("Killing applications: %s", applicationIDs));
+    ExecutorService executor = Executors.newSingleThreadExecutor();
 
     for (final String appId : applicationIDs) {
+      Future<?> future = executor.submit(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            YarnUtils.killAppOnCluster(yarnClient, appId, log);
+          } catch (final Throwable t) {
+            log.warn("something happened while trying to kill this job: " + appId, t);
+          }
+        }
+      });
+      // wait for the kill with a timeout
       try {
-        YarnUtils.killAppOnCluster(yarnClient, appId, log);
-      } catch (final Throwable t) {
-        log.warn("something happened while trying to kill this job: " + appId, t);
+        future.get(YARN_APP_TIMEOUT_IN_MILLIONSECONDS, TimeUnit.MILLISECONDS);
```

Comment:
Can we consider adding a unit test for the killAllAppscluster method?

Reply:
I couldn't find a way to make a unit test work: mock sleeping in unit test is hard, instead I tested in an executor pod, the killing works well and the future timed out after 5 seconds